### DISCUSSION TITLE
Support for ESM modules

### DISCUSSION THREAD

**Discussion by loklaan** [keystonejs/keystone#6647] [Feature requests]
Hey folks!

It looks like some commonly used packages (read: [Sindres Sorhus's pkgs](https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c)) are recently making the switch to ESM-only exports, and at the moment they won't work with the keystone cli, since under the hood isn't setup for that.

I'd wager the Keystone team have a plan here, since the forward-incompatibility is a big looming issue for _most_ kinds of JS project, but I'd like put a bookmark on the topic here!

I've been tracking similar discussions across:
- [Gatsby](https://github.com/gatsbyjs/gatsby/discussions/31599)
- [NextJS](https://github.com/vercel/next.js/discussions/27876)

---

_This was a question I asked originally [in Slack](https://keystonejs.slack.com/archives/C01STDMEW3S/p1632377336422400), but afterwards I thought it would be nice to track this discussion in Github. Thanks again @raveling for the quick response there!_ 👋
> _Unfortunately this isn't currently possible. Next.js has experimental support for Node ESM, but it's not ready for prime time yet. We may be able to provide a better story for this use case after it's production ready. Hope this helps!_
> _- raveling_

**Comment by raveling** [comment]
Thanks for posting this issue @loklaan 🙂

**Comment by raveling** [comment]
@loklaan we will support this eventually. Hope this helps 🙂

**Comment by wingedrhino** [comment]
Is there a rough ETA on when keystonejs can support ESM? I am evaluating it for a new project and I have a few months before it hits production. Some of my dependencies are ESM-only.

**Comment by mmachatschek** [comment]
@dcousens could we get the current on this topic?

What would it take to make keystone compatible for both ESM and CJS?

**Reply by mmachatschek** [reply]
One blocker is the current implementation of the config generator that builds `.keystone/config.js` in CJS format (which doesn't allow ESM only imports of packages like `lodash-es`, `@sindresorhus/slugify`)

**Reply by dcousens** [reply]
@mmachatschek in effect, if we can work on making the CI pass for https://github.com/keystonejs/keystone/pull/8820 - we'll be some of the way there

**Comment by learnwithus** [comment]
Very much looking forward to ESM support.  I have a few dependencies that I'm unable to upgrade because they've gone ESM only.

**Reply by unrevised6419** [reply]
From my understanding, the status of ESM for KS is a bit tricky now in its current state.

Also, this repo uses preconstruct
 
https://github.com/preconstruct/preconstruct/issues/513#issuecomment-1529157753
https://github.com/preconstruct/preconstruct/issues/471
